### DISCUSSION TITLE
Is docker-compose AIRFLOW_UID env variable confusing?

### DISCUSSION THREAD

**Discussion by bsdz** [apache/airflow#15579] [Q&A] [answered]
Hi

I've been tinkering around with the docker-compose instructions [here](https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html?highlight=docker%20compose).

While customising the compose file (to support adding the airflow user to a docker group to allow access to /var/run/docker.sock) I became a little confused with the compose file's AIRFLOW_UID environment variable. The name suggests it's the UID for the "airflow" user within the containers. However, that's not possible as AIRFLOW_UID is set as a build argument for the upstream image on dockerhub.

What appears to happen instead is the USER Dockerfile instruction is overridden & a new "default" user is added to passwd file using the environment variable AIRFLOW_UID.

For example, see script below: 

```bash
foo@host:~/airflow-compose-test$ curl -LfO 'https://airflow.apache.org/docs/apache-airflow/2.0.2/docker-compose.yaml'
foo@host:~/airflow-compose-test$ mkdir ./dags ./logs ./plugins
foo@host:~/airflow-compose-test$ echo -e "AIRFLOW_UID=$(id -u)\nAIRFLOW_GID=0" > .env
foo@host:~/airflow-compose-test$ cat .env 
AIRFLOW_UID=1000
AIRFLOW_GID=0
foo@host:~/airflow-compose-test$ docker-compose up airflow-init
...
airflow-init_1       | Admin user airflow created
airflow-init_1       | 2.0.2
airflow-compose-test_airflow-init_1 exited with code 0
foo@host:~/airflow-compose-test$ docker-compose run airflow-worker bash
Creating airflow-compose-test_airflow-worker_run ... done
BACKEND=postgresql+psycopg2
DB_HOST=postgres
DB_PORT=5432

default@a380ba45b47f:/opt/airflow$ id
uid=1000(default) gid=0(root) groups=0(root)
default@a380ba45b47f:/opt/airflow$ cat /etc/passwd
...
airflow:x:50000:0:,,,:/home/airflow:/bin/bash
default:x:1000:0:default user:/home/airflow:/sbin/nologin
default@a380ba45b47f:/opt/airflow$ groups
root
```

Is that intentional? If it is, would it perhaps be better to rename AIRFLOW_UID in the docker-compose to another name like AIRFLOW_HOST_UID or better? Admittedly, I might be missing something too.

Kind regs
Blair

**Comment by potiuk** [comment, marked as answer]
Yes. It is intentional. It is to allow to run arbitrary UID to run the image and make it OpenShift compatible. It is expleined in detain in the documentation https://airflow.apache.org/docs/docker-stack/entrypoint.html#allowing-arbitrary-user-to-run-the-container

**Reply by bsdz** [reply]
I did read that section (and the relevant parts in the openshift document) and have to admit I still find it confusing. AIRFLOW_UID in the Dockerfile appears to be a different thing to AIRFLOW_UID in the docker-compose.

**Reply by potiuk** [reply]
It's the same and i re-read it - as I understand it, it answers all your questions and doubts from the above questiob, but maybe I  have some assumptions that you don'y. Could you please explain why do you think they are different ? 

This section describes exactly why (OpenShift Compatibility), why (it explains that new user is created and why it is needed) and gives you guidelines on how to use it (GID = 0). What's even more in case you make mistake, and use.differnt guide, the entry point in airflow 2.0 will fail with helpful message. 

I would really love (since airflow is community managed project) if you could explain why this is confusing and possibly even come up with some fixes to documentation that could make it clearer. I find that confused users (mostly because of lack of assumptions of those who wrote the documentation) are best to mke PRs that mąkę the documentation better.

 Could you please elaboratów on what is do confusion and maybe after the discussion you could come up with PR improving the doc ? You are most welcome to do it .

**Reply by bsdz** [reply]
UPDATE: By mistake (editing from the phone) @potiuk edited it instead of replied to. This is the original text:

>  It's the same and i re-read it - as I understand it, it answers all your questions and doubts from the above questiob, but maybe I have some assumptions that you don'y. Could you please explain why do you think they are different ?
 Well for a start it doesn't even mention AIRFLOW_UID so how can it answer all my questions. Just to requote:

 "If it is (intentional), would it perhaps be better to rename AIRFLOW_UID in the docker-compose to another name like AIRFLOW_HOST_UID or better?"

> This section describes exactly why (OpenShift Compatibility), why (it explains that new user is created and why it is needed) and gives you guidelines on how to use it (GID = 0). What's even more in case you make mistake, and use.differnt guide, the entry point in airflow 2.0 will fail with helpful message.

Yeah got the group-id technique to share resources. IMHO that's all the openshift doc really talks about.

> I would really love (since airflow is community managed project) if you could explain why this is confusing and possibly even come up with some fixes to documentation that could make it clearer. I find that confused users (mostly because of lack of assumptions of those who wrote the documentation) are best to mke PRs that mąkę the documentation better.

> Could you please elaboratów on what is do confusion and maybe after the discussion you could come up with PR improving the doc ? You are most welcome to do it .

I would love to contribute however the problem I am having right now is 1) you don't believe me when I say something is confusing 2) i have provided an example and you still don't believe me 3) what is there to fix if it's not confusing (like you suggest)? 4) I felt somewhat patronised by your response... (perhaps not intentional..)

**Reply by bsdz** [reply]
Why did you edit my comment with your response? You only need to reply and quote me, not edit my comment. 

In any case, thanks for your reply. Here's a picture that hopefully makes what I find confusing clearer to you..

![image](https://user-images.githubusercontent.com/6608130/116549339-3da6f700-a8ed-11eb-91b9-8c06805f5ac7.png)

Regarding your reply to my 4th point: "i have completely no influence on what feelings my words cause". You clearly do have influence on what feelings your words cause; after all, you choose what words you use and how to put them together.

**Reply by potiuk** [reply]
> Why did you edit my comment with your response? You only need to reply and quote me, not edit my comment.

Ough. sorry. I was doing that on my phone, pure mistake. I thought I was doing "quote reply" :). I will revert and make proper comment shortly.

**Reply by potiuk** [reply]
Well for a start it doesn't even mention AIRFLOW_UID so how can it answer all my questions. 

Ok. That's s a very concrete and good comment. This is exactly what I mean by assumptions that I have that the reader might not. are you willing to submit the change to add it or would you prefer for me.to add it ? As mentioned before Airflow is community work and many people are contributing. I really value users contributing documentation. You can find many cases in the past commits when it happens and our documentation got way better thanks to it. 

Just let me know if you would like to add.it 

> Just to requote:

"_If it is (intentional), would it perhaps be better to rename **AIRFLOW_UID** in the docker-compose to another name like AIRFLOW_HOST_UID or better?_"

Oh yeah. I missed that comment, really sorry. Somehow I skipped over i  it is a bad idea.it would  be even more confusing because it is not HOST_UID. AIARFLOW_UID  is the very same user id that can be set via USER  command in Dockerfile. This is a common technique and the corresponding parameters in both docker compose a docker comma nd line are named uid/gid so prefixing it with host would be totally confusing. We are already using HOST_UID and HOST_GiD in our test harness (but for different reason - we are passing the host uid/gid to within container so that it can fix ownership of files created inside the container). But this has nothing to do with the uid/gid used when you enter the container. 

> Yeah got the group-id technique to share resources. IMHO that's all the openshift doc really talks about. 

Not really. The section is about the fact that OpenShift overrides the user run in the container with a random, arbitrary user id. Which is precisely what the AIRFLOW UID is about. As mentioned above, i am happy to add specific reference to the AiRFLOW_UID if you do not feel like it.


> Could you please elaboratów on what is do confusion and maybe after the discussion you could come up with PR improving the doc ? You are most welcome to do it .

I would love to contribute however the problem I am having right now is 1) you don't believe me when I say something is confusing 2) i have provided an example and you still don't believe me 3) what is there to fix if it's not confusing (like you suggest)? 4) I felt somewhat patronised by your response... (perhaps not intentional..)

1) i believe you. But i had simply no idea what was wrong. It was all clear to me - precisely because I have my own assumptions which you don't. And I ask you to clarify this. No more, no less. I also invited you to update the documentation after we agree what could be corrected. Until you mentioned that AIRFLOW_UID is missing in the other part of documentation i simply had no idea it could be confusing because it was obvious to me. And since airflow (again) is a community driven project - this is just normal to ask people who raise their concern to.provide a fix. This is how our community works and we value such contributions enormously. 

2) i missed the question, sorry but it is a bad idea as explained above.

3) that was my question :). I had no idea what was wrong and asked you to clarify 

4) sorry if you felt like that, i have completely no influence on what feelings my words cause, this is your personal reception of it. but i know.what my intentions were. they were certainly not patronizing but trying to be helpful, dig deeper and invite you to become a.contributor. no more, no less.

So i would really love to know -  do you think there is anything else confusing beside lack of reference to that specific parameter?

**Reply by potiuk** [reply]
> In any case, thanks for your reply. Here's a picture that hopefully makes what I find confusing clearer to you..

Ok. I willl make a proposal of PR to explain it better then. I will ask you to review and see if this explains better and removes the confusion

**Reply by potiuk** [reply]
> Regarding your reply to my 4th point: "i have completely no influence on what feelings my words cause". You clearly do have influence on what feelings your words cause; after all, you choose what words you use and how to put them together.

Your feelings are yours, not mine. My intentions were clear, if you mistunderstood them, apologies (again) but again, I have no influence on your feelings. There are some cultural and personal differences, and different people have different sensitivity. How other people will react on your words (especially written and when you do not know them) is something totally out of your control or knowledge). 

Usually when someone apologises for unintentional misunderstanding of their words, and explains what the intentions are, that ends the discussion and there is no need to call it out again, - "assume best intentions" is important in the community.

**Reply by bsdz** [reply]
> > In any case, thanks for your reply. Here's a picture that hopefully makes what I find confusing clearer to you..
> 
> Ok. I willl make a proposal of PR to explain it better then. I will ask you to review and see if this explains better and removes the confusion

Cool.

**Reply by bsdz** [reply]
> > Regarding your reply to my 4th point: "i have completely no influence on what feelings my words cause". You clearly do have influence on what feelings your words cause; after all, you choose what words you use and how to put them together.
> 
> Your feelings are yours, not mine. My intentions were clear, if you mistunderstood them, apologies (again) but again, I have no influence on your feelings. There are some cultural and personal differences, and different people have different sensitivity. How other people will react on your words (especially written and when you do not know them) is something totally out of your control or knowledge).
> 
> Usually when someone apologises for unintentional misunderstanding of their words, and explains what the intentions are, that ends the discussion and there is no need to call it out again, - "assume best intentions" is important in the community.

No worries

**Reply by potiuk** [reply]
PR Created: https://github.com/apache/airflow/pull/15592 -> Pls take a look, if this is fess confusing. Please note that there is a "suggest" button so you might directly add suggestions if you want to make a suggestion.

**Reply by potiuk** [reply]
@bsdz  -> is the new description in #15592 clearer? Does it remove all the confusion? Any more points to add ?

**Reply by bsdz** [reply]
Yes, I added a suggestion. Forgot to press submit-review button yesterday. Thanks.
